scripts/publish.ts

```diff
@@ -0,0 +1,408 @@
+          if (parsedPackageVersion === null) {
+            throw new CommandError(
+              formatInvalidPackageVersionError(packageName, packageVersion),
+            );
+          }
+          const highestPublishedPrereleaseNumber =
+            findHighestPublishedPrereleaseNumber(
+              [...(packageNameToVersionHistory.get(packageName) ?? [])],
+              parsedPackageVersion,
+              prereleaseIdentifier,
+            );
+          const incrementedPackageVersion = incrementVersionByPrerelease(
+            parsedPackageVersion,
+            prereleaseIdentifier,
+            highestPublishedPrereleaseNumber,
+          );
+          return [packageName, incrementedPackageVersion] as const;
+        }),
+      ),
+    );
+  }
+  function formatInvalidPackageVersionError(
+    packageName: string,
+    packageVersion: string,
+  ): string {
+    return `Package \`${packageName}\` has an invalid version \`${packageVersion}\` in its manifest.`;
+  }
+
+  // Ensure the new versions aren't already in use.
+  if (!args["tolerate-republish"]) {
+    packageNameToReleaseVersion.forEach(
+      (packageReleaseVersion, packageName) => {
+        const packageVersionHistory =
+          packageNameToVersionHistory.get(packageName);
+        if (packageVersionHistory?.has(packageReleaseVersion)) {
+          throw new CommandError(
+            `Package \`${packageName}@${packageReleaseVersion}\` has already been published.`,
+          );
+        }
+      },
+    );
+  }
+
+  // Update the manifest of each package that is part of the release.
+  console.log(`✏️  Updating manifest${s}...`);
+  for (const nonPrivateWorkspaceName of nonPrivateWorkspaceNames) {
+    const nonPrivateWorkspaceReleaseVersion = packageNameToReleaseVersion.get(
+      nonPrivateWorkspaceName,
+    );
+    await exec(
+      "yarn",
+      [
+        ...["workspace", nonPrivateWorkspaceName],
+        ...["version", nonPrivateWorkspaceReleaseVersion ?? "decline"],
+        "--deferred",
+      ],
+      { dryRun },
+    );
+  }
+  await exec("yarn", ["version", "apply", "--all"], { dryRun });
+
+  // If running in stable mode, both commit and tag the changes. Otherwise, only
+  // tag them.
+  console.log(`🏷️  ${mode === "stable" ? "Committing" : "Tagging"} changes...`);
+  const { stdout: gitDiff } = await exec("git", ["diff", "--name-only"]);
+  if (gitDiff.length === 0 && !dryRun) {
+    throw new CommandError(
+      "There are no changed package versions to commit or tag.",
+    );
+  } else if (mode === "stable") {
+    const commitMessage = `release: ${packageNameToReleaseVersion.size} new package${s}`;
+    await exec("git", ["commit", "--all", "--message", commitMessage], {
+      dryRun,
+    });
+  }
+  for (const [
+    packageName,
+    packageReleaseVersion,
+  ] of packageNameToReleaseVersion) {
+    const gitTagName = `${packageName}@${packageReleaseVersion}`;
+    await exec(
```

Comment:
Suggestion for a future version: it could be nice to have a `--verbose`/`-v` mode where `exec` logs the command.